[8.13][SecuritySolution][Endpoint] Retrieve (API) agent status for SentinelOne agents and display it in the UI

paul-tavares opened the pull request:
## Summary

- New internal API that returns the agent status for one of the supported agent types (currently `endpoint` or `sentinel_one`) - route: `/internal/api/endpoint/agent_status`
    - Supported query params: `agentType` and `agentIds`
    - The response will include a record for each agent ID that was requested. Each record contains a `found: boolean` that indicates whether the agent id information was found or not (could be the case when agents are un-enrolled)
    - NOTE: currently, this API will return only statuses for SentinelOne in 8.13. In the next release, it will be expanded with a framework that enables it to return status for any supported `agentType`


__________

API Request: `GET /internal/api/endpoint/agent_status?agentType=sentinel_one&agentIds=1111&agentIds=2222agentIds=invalid-id`

API Response:

```json5
{
  "data": {
    "1111": {
      "agentType": "sentinel_one",
      "id": "1111",
      "found": true,
      "status": "healthy",
      "isolated": true, // <<<< Note host is isolated
      "lastSeen": "2024-02-07T19:03:31.287300Z",
      "pendingActions": {
        "execute": 0,
        "upload": 0,
        "unisolate": 0,
        "isolate": 0,
        "get-file": 0,
        "kill-process": 0,
        "suspend-process": 0,
        "running-processes": 0
      }
    },
    "2222": {
      "agentType": "sentinel_one",
      "id": "2222",
      "found": true,
      "status": "healthy",
      "isolated": false,
      "lastSeen": "2024-02-07T14:45:02.830739Z",
      "pendingActions": {
        "execute": 0,
        "upload": 0,
        "unisolate": 0,
        "isolate": 1, // <<< Note: host is Isolating
        "get-file": 0,
        "kill-process": 0,
        "suspend-process": 0,
        "running-processes": 0
      }
    }
  },
  {
    "invalid-id": {
      "agentType": "sentinel_one",
      "id": "invalid",
      "found": false,
      "status": "unenrolled",
      "isolated": false,
      "lastSeen": "",
      "pendingActions": {
        "execute": 0,
        "upload": 0,
        "unisolate": 0,
        "isolate": 0,
        "get-file": 0,
        "kill-process": 0,
        "suspend-process": 0,
        "running-processes": 0
      }
    }
}
```




### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

paul-tavares committed:
new API route + schemas + type (no response yet)

paul-tavares committed:
service that returns SentinelOne agent status

paul-tavares committed:
set `uuids` to a string (join of agent ids)

ashokaditya commented on x-pack/plugins/security_solution/common/api/endpoint/agent/get_agent_status_route.ts:
```diff
@@ -0,0 +1,41 @@
+      AgentStatusAgentIdSchema,
+    ]),
+
+    agentType: schema.maybe(
```
This would limit the agent details/status search to one agent type. We should make this so it accepts multiple agent types.

paul-tavares replied:
Adding support for multiple agent types would likely need to convert this API to a `POST` with body and then have an array of input with `{ agentType: string; agentId: string }` - making the API much more complex and perhaps harder to work with. Lets hold off on that until the need arises. My thinking is that the caller would issue two api calls to get the status, but perhaps your suggestion would work best when we start to introduce support for many external EDR systems. I think the real need will come when we start to show a list of Agents/Hosts from multiple EDR system in one view (similar to our Endpoint list) where we need to show the status.

paul-tavares committed:
Rename file (`mock` to `mocks`)

paul-tavares committed:
tests for getAgentStatus service

paul-tavares committed:
tests for agent status API route handler

paul-tavares committed:
tests for agent status route schema

paul-tavares commented:
/ci

ashokaditya committed:
API changes to include uninstalled status

We need this to enable/disable isolate/release actions. It should be disabled on agents that are uninstalled or pending uninstalled.

ashokaditya committed:
Use the internal API to fetch sentinel one info

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

ashokaditya commented on x-pack/plugins/security_solution/public/detections/components/host_isolation/use_sentinelone_host_isolation.tsx:
```diff
@@ -6,45 +6,49 @@
+    enabled: !(
+      sentinelOneManualHostActionsEnabled &&
+      isEmpty(agentIds.filter((agentId) => agentId.trim().length))
+    ),
```
note: Needs a `refetchInterval` value from options that should be modified from Responder view.

ashokaditya committed:
unify agentType schema definitions

ashokaditya committed:
Merge branch 'main' into task/olm-8246-new-agent-status-api

ashokaditya committed:
fix test

ashokaditya committed:
Merge branch 'main' into task/olm-8246-new-agent-status-api

paul-tavares reviewed (commented):
@ashokaditya ,
Thanks for all the changes to the UI. did a quick review and looks good to me. FYI: I changed the title of the PR

ashokaditya committed:
use lastSeen from API

PhilippeOberti reviewed (approved):
LGTM for the Threat Hunting Investigations team!

ashokaditya committed:
fix type

ashokaditya committed:
Merge branch 'main' into task/olm-8246-new-agent-status-api
